include/libtwofloat/arithmetics/double-word-arithmetic.hpp

```diff
@@ -233,5 +233,628 @@ inline two<T> div(const two<T> &x, const two<T> &y) {
+
+static const float fp32_inv_fact[n_inv_fact][2] = {
+  { 1.6666666e-01,  9.2518585e-18},
+  { 4.1666666e-02,  2.3129646e-18},
+  { 8.3333333e-03,  1.1564823e-19},
+  { 1.3888888e-03, -5.3005439e-20},
+  { 1.9841269e-04,  1.7209558e-22},
+  { 2.4801587e-05,  2.1511947e-23},
+  { 2.7557319e-06, -1.8583932e-22},
+  { 2.7557319e-07,  2.3767714e-23},
+  { 2.5052108e-08, -1.4488140e-24},
+  { 2.0876756e-09, -1.2073450e-25},
+  { 1.6059043e-10,  1.2585294e-26},
+  { 1.1470745e-11,  2.0655512e-28},
+  { 7.6471637e-13,  7.0387287e-30},
+  { 4.7794773e-14,  4.3992054e-31},
+  { 2.8114572e-15,  1.6508842e-31}
+};
+
+static const double fp64_inv_fact[n_inv_fact][2] = {
+  { 1.66666666666666657e-01,  9.25185853854297066e-18},
+  { 4.16666666666666644e-02,  2.31296463463574266e-18},
+  { 8.33333333333333322e-03,  1.15648231731787138e-19},
+  { 1.38888888888888894e-03, -5.30054395437357706e-20},
+  { 1.98412698412698413e-04,  1.72095582934207053e-22},
+  { 2.48015873015873016e-05,  2.15119478667758816e-23},
+  { 2.75573192239858925e-06, -1.85839327404647208e-22},
+  { 2.75573192239858883e-07,  2.37677146222502973e-23},
+  { 2.50521083854417202e-08, -1.44881407093591197e-24},
+  { 2.08767569878681002e-09, -1.20734505911325997e-25},
+  { 1.60590438368216133e-10,  1.25852945887520981e-26},
+  { 1.14707455977297245e-11,  2.06555127528307454e-28},
+  { 7.64716373181981641e-13,  7.03872877733453001e-30},
+  { 4.77947733238738525e-14,  4.39920548583408126e-31},
+  { 2.81145725434552060e-15,  1.65088427308614326e-31}
+};
+
+// Reference: QD / dd_real.cpp
+/* Table of sin(k * pi/16) and cos(k * pi/16). */
+static const float fp32_cos_table [4][2] = {
+  {9.8078528e-01, 1.8546939e-17},
+  {9.2387953e-01, 1.7645047e-17},
+  {8.3146961e-01, 1.4073856e-18},
+  {7.0710678e-01, -4.8336466e-17}
+};
+static const float fp32_sin_table [4][2] = {
+  {1.9509032e-01, -7.9910790e-18},
+  {3.8268343e-01, -1.0050772e-17},
+  {5.5557023e-01,  4.7094109e-17},
+  {7.0710678e-01, -4.8336466e-17}
+};
+
+static const double fp64_cos_table [4][2] = {
+  {9.807852804032304306e-01, 1.854693999782500573e-17},
+  {9.238795325112867385e-01, 1.764504708433667706e-17},
+  {8.314696123025452357e-01, 1.407385698472802389e-18},
+  {7.071067811865475727e-01, -4.833646656726456726e-17}
+};
+static const double fp64_sin_table [4][2] = {
+  {1.950903220161282758e-01, -7.991079068461731263e-18},
+  {3.826834323650897818e-01, -1.005077269646158761e-17},
+  {5.555702330196021776e-01,  4.709410940561676821e-17},
+  {7.071067811865475727e-01, -4.833646656726456726e-17}
+};
+
+// Notice the definition of the qd namespace
+namespace qd {
+
+  // Reference: QD / inline.h
+  /* Computes the nearest integer to input. */
+  template <typename T>
+  inline T nint(T input) {
+    T pointfive;
+
+    if constexpr (std::is_same_v<T, float>) {
+      pointfive = 0.5f;
+    } else if constexpr (std::is_same_v<T, double>) {
+      pointfive = 0.5;
+    } else {
+      std::cerr << "Other types not supported\n";
+    }
```

Comment:
Depending on the rounding mode that the processor supports this can be done in a single instruction. We need some way for the user to choose the implementation here too.

Reply:
Check above https://github.com/esa-tu-darmstadt/twofloat/pull/1#discussion_r1481692804.